### ISSUE TITLE
[ruler] unbounded growth in group evaluation latency after last rule is deleted

### ISSUE THREAD

**Issue by nicolai86** [cortexproject/cortex#3359]
**Describe the bug**
When the very last rule is deleted the ruler enters a state where the group evaluation latency is growing unbounded. Once this state is entered new rules are not picked up, and the only way to get the rulers to evaluate any rules again is to restart the rulers.

**To Reproduce**
Steps to reproduce the behaviour:

1. Start Cortex, with a rulers and an alertmanager. I used 3 rulers with memberlist and 3 alertmanagers in HA configuration, but it really doesn't matter.

2. Create tenant alert configuration, ie this one:
```
> POST /api/v1/alerts HTTP/1.1
Host: 127.0.0.1:8080
Content-Type: application/yaml
X-Scope-Orgid: tenant-0

---
alertmanager_config: |
  global:
    resolve_timeout: 5s
  route:
    receiver: webhook
    group_interval: 1m
  receivers:
    - name: webhook
      webhook_configs:
      - send_resolved: false
        url: http://alertmanager-sns-forwarder-svc.cortex:9087/alert/us-west-2/611128412359/test

HTTP/1.1 201 Created
Content-Length: 0
Date: Sat, 17 Oct 2020 02:10:30 GMT
```
Note that I've port-forward the alertmanager to localhost:8080 in this example

3. create ruler configuration for tenant:
```
POST /api/prom/rules/alarms-0-000 HTTP/1.1
Host: 127.0.0.1:8081
Content-Type: application/yaml
X-Scope-Orgid: tenant-0

name: "alarms-1-001"
interval: 60s
rules:
- alert: alarm-1
  annotations:
    message: test 1
  expr: vector(1)
  labels:
    severity: none

< HTTP/1.1 202 Accepted
Content-Length: 58
Content-Type: application/json
Date: Sat, 17 Oct 2020 02:10:30 GMT

{"status":"success","data":null,"errorType":"","error":""}
```
Note that I've port-forwarded the ruler to localhost:8081 in this example
4. wait for rule to be evaluated at least once.
5. delete rule config
```
> DELETE /api/prom/rules/alarms-0-000 HTTP/1.1
Host: 127.0.0.1:8081
Content-Type: application/yaml
X-Scope-Orgid: tenant-0

< HTTP/1.1 202 Accepted
Content-Length: 58
Content-Type: application/json
Date: Sat, 17 Oct 2020 02:14:59 GMT
{"status":"success","data":null,"errorType":"","error":""}
```
6. delete alertmanager config

```
> DELETE /api/v1/alerts HTTP/1.1
Host: 127.0.0.1:8080
X-Scope-Orgid: tenant-0

< HTTP/1.1 200 OK
Date: Sat, 17 Oct 2020 02:15:00 GMT
Content-Length: 0
```
7. wait until files are deleted from ruler/ alertmanager store
8. follow step 2. and 3. to create alertmanager and rule configuration again

**Expected behavior**
The ruler should evaluate the rules again. However, the rules are not picked up, and the ruler doesn't perform any action.
[Example 1](https://snapshot.raintank.io/dashboard/snapshot/8381v1WkWs0i6w1obRNmxW9nFu2rrI6m) [Example 2, after 26h](https://snapshot.raintank.io/dashboard/snapshot/qRvSrKPtqv2tDvH3f5TCFFoU8un4U1xu)

**Environment:**
 - Infrastructure: EKS
 - Deployment tool: helm

**Storage Engine**
- [x] Blocks
- [ ] Chunks

**Additional Context**

alertmanager configuration
```
alertmanager:
  data_dir: /data/alertmanager
  external_url: /alertmanager
  cluster_bind_address: "0.0.0.0:9094"
  storage:
    type: s3
    s3:
      s3: s3://us-west-2/<some-bucket>
  enable_api: true
```

ruler configuration:
```
ruler:
  storage:
    type: s3
    s3:
      s3: s3://us-west-2/<another-bucket>
  ring:
    kvstore:
      store: memberlist
      prefix: ruler/
  alertmanager_url: http://alertmanager.cortex.svc.cluster.local/alertmanager/
  enable_sharding: true
  enable_api: true
```

memberlist
```
memberlist:
  bind_port: 7946
  join_members:
    - ingester-memberlist.cortex.svc.cluster.local:7946
    - store-gateway-memberlist.cortex.svc.cluster.local:7946
    - distributor-memberlist.cortex.svc.cluster.local:7946
    - compactor-memberlist.cortex.svc.cluster.local:7946
  abort_if_cluster_join_fails: false
```

**Comment by gotjosh** [comment]
Thanks for reporting this @nicolai86, I have it on my agenda to take a look at it this week 😄

**Comment by gotjosh** [comment]
Hey @nicolai86,

I've been trying to reproduce this during the last few days without much luck - I have tried several combinations of rules/groups, alongside deletion/creation of configuration timings. 

The one thing I am yet to try is using memberlist for the ring but according to the issue, it does not matter?

Can you please provide me with a bit more detail that can help me replicate? I have noticed you're using helm to deploy - are these [our own helm charts](https://github.com/cortexproject/cortex-helm-chart)?

**Comment by gotjosh** [comment]
Also, just to be sure - can you please confirm how you're evaluating latency here? It should be:

```
rate(cortex_prometheus_rule_group_duration_seconds_sum[5m])
  /
rate(cortex_prometheus_rule_group_duration_seconds_count[5m])
```

Then, could you verify what the [Prometheus compatible endpoint](https://cortexmetrics.io/docs/api/#list-rules) returns before and after.

**Comment by nicolai86** [comment]
sorry for the delay I moved so didn't have time to look into this.

> The one thing I am yet to try is using memberlist for the ring but according to the issue, it does not matter?

we're using memberlist - I never checked if this works with other kv store mechanisms.

> I have noticed you're using helm to deploy

we're using a private fork of the helm chart with a couple of changes to it. I'll see if i can post a copy of it in this issue. 

maybe relevant - the version of cortex we are using is cortex at bbf01e830.

```
rate(cortex_prometheus_rule_group_duration_seconds_sum[5m])
  /
rate(cortex_prometheus_rule_group_duration_seconds_count[5m])
```

this is NaN, as both values are 0.

The dashboard we're using comes from [weaveworks/cortex-dashboards](https://github.com/weaveworks/cortex-dashboards/blob/edde28bd6950557fdfe45cdd3df0de7f213f7a18/cortex-ruler.dashboard.py#L82) and uses a different metric: `cortex_group_evaluation_latency_seconds_bucket` 

Maybe that's the issue?

**Comment by gotjosh** [comment]
> this is NaN, as both values are 0.

This would make sense, if the rules are not being picked up. 

> The dashboard we're using comes from weaveworks/cortex-dashboards and uses a different metric: cortex_group_evaluation_latency_seconds_bucket
Maybe that's the issue?

It's hard to tell based on the graphs without data, at a glance I see that the query of the graph that growths unbounded is `max(time()-(cortex_prometheus_rule_group_last_evaluation_timestamp_seconds>0))*1e3` which makes sense in the case where `cortex_prometheus_rule_group_last_evaluation_timestamp_seconds` is 0. It'll always pick the value of `time()`.

I'm puzzled by the fact that rules are never picked up again, during my tests this never happened no matter which circumstances - as soon as I created new rules after deleting they'll get evaluated. 

I don't think the commit of Cortex is relevant as this area has not seen significant changes since. Could you please post the values you see of during the configuration changes `cortex_prometheus_rule_group_last_evaluation_timestamp_seconds` - I'm more surprised there's nothing in the logs.

**Comment by nicolai86** [comment]
I synced with @gotjosh today - we were able to reproduce the issue with the version I mentioned above. However, upgrading to 1.5.0 fixes the issue.

@gotjosh pointed towards a refactoring which was introduced in later commits which indeed resolves this issue.

**Comment by gotjosh** [comment]
Thanks, @nicolai86 !

I've managed to pin-point the exact problem after that last round of testing. The fix was introduced as part of [this pull request](https://github.com/cortexproject/cortex/pull/3235/files#diff-dbc64ce58eaf88369c34a86fb77a822764408eca8a0950f378c029a9f8f5c734). 

Just like I suspected, before this commit, we didn't verify whenever or not a user already had a manager running if their rules were the same as the copy we keep on-disk for checksum verification and loading.
